### PR TITLE
Signup

### PR DESCRIPTION
**Pull request by Rodrigo-Miranda-0**
This PR creates many of the components used through the landing page of the Target site.
It also creates logic elements such as:

- Router

- API request handler

Finally, it utilizes all previously mentioned items to create the `Signup` page

The `VITE_API_URL` environment variable is the current Heroku API server, and its value can be found [here](https://www.notion.so/rootstrap/Target-1c963bd0c09345668f4eb34ab4b5271a)

Screenshot 2024-01-26 at 15 15 46

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (eugenialcaraz) [inline .prettierrc:1]: Will you have a setup here?

File: src/App.tsx [hunk 2]
```diff
@@ -1,35 +1,16 @@
...
-        Click on the Vite and React logos to learn more
-      </p>
-    </>
-  )
+    <BrowserRouter>
+      <Routes>
+        <Route path="/" element={<Navigate to="/signup" />} />
+        <Route path="/signup" element={<SignupPage />} />
+      </Routes>
+    </BrowserRouter>
```
Comment (eugenialcaraz) [inline src/App.tsx]: Most probably you will add more routes in here, and the code can get a little repetitive. However you could create an array of routes and map through it. Something like this:

```suggestion

// before the return: 
//const routes = [
// {path:'/', element: <Navigate to='/signup'/>},
// {path:'/signup' element:<SignupPage/>}];

    <BrowserRouter>
      <Routes>
      {routes.map(({path,element}) => <Route path={path} element={element}/>)}
      </Routes>
    </BrowserRouter>
```

Comment (eugenialcaraz) [reply src/App.tsx]: Another suggestion is to use constants for all the strings you type manually across the whole app.
For example here you typed twice `'/signup'`, but if you have a file of constants and you import it from there,  typing mistakes wont break the app, and if in the future the client asks to change it to "create account" you just change it in one place

File: src/components/common/ButtonComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { MouseEventHandler } from "react";
+
+interface ButtonComponentProps {
+  className?: string;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
+  children: React.ReactNode;
+  disabled: boolean;
+}
+const ButtonComponent = (props: ButtonComponentProps) => {
```
Comment (eugenialcaraz) [inline src/components/common/ButtonComponent.tsx]: You can destruct props in the component, so you don't have to use `props.className`, `props.onClick`, etc.

```suggestion
const ButtonComponent = ({ className, onClick, children, disabled }: ButtonComponentProps) => {
```

File: src/components/common/ButtonComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  return (
+    <button
+      className={`${
+        props.className ? props.className : ""
+      } text-white text-center text-xs font-semibold bg-black self-center w-[185px] max-w-full justify-center items-center mt-6 px-16 py-4 max-md:px-5 disabled:opacity-30`}
```
Comment (eugenialcaraz) [inline src/components/common/ButtonComponent.tsx]: You can add a default value of the className to be `""`, then you avoid the conditional. 
Another thing, the way you have it here, the `className` received will be overwritten by the fixed classes you added. If you want it to be the other way round, the received className needs to be at the end.

Like this the component classes will override the received `className`
```suggestion

//at the definition of the component: const ButtonComponent = ({ className = "", onClick, children, disabled }: //ButtonComponentProps) => {

      className={`${
        props.className
      } text-white text-center text-xs font-semibold bg-black self-center w-[185px] max-w-full justify-center items-center mt-6 px-16 py-4 max-md:px-5 disabled:opacity-30`}
```


Like this the received classes will override:

```suggestion
className={`text-white text-center text-xs font-semibold bg-black self-center w-[185px] max-w-full justify-center items-center mt-6 px-16 py-4 max-md:px-5 disabled:opacity-30 ${props.className} `}
```

File: src/components/common/LeftMainPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { ReactElement } from "react";
+import HamburguerMenu from "./HamburgerMenu";
+
+const LeftMainPanel = (props: { children: ReactElement }) => {
```
Comment (eugenialcaraz) [inline src/components/common/LeftMainPanel.tsx:4]: I think this works and `ts` doesn't scream:

```suggestion
const LeftMainPanel = ({ children}: ReactElement) => {
```

File: src/components/common/MainLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import LeftMainPanel from "./LeftMainPanel";
+import RightMainPanel from "./RightMainPanel";
+import { ReactElement } from "react";
+
+const MainLayout = (props: { children: ReactElement }) => {
```
Comment (eugenialcaraz) [inline src/components/common/MainLayout.tsx:5]: Same here:
```suggestion
const MainLayout = ({ children}: ReactElement) => {
```

File: src/components/common/HamburgerMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const HamburguerMenu = () => {
+  return (
+    <img
+      loading="lazy"
+      src="src/assets/menu.svg"
```
Comment (eugenialcaraz) [inline src/components/common/HamburgerMenu.tsx:5]: it's a good practice to import the src as an element
```suggestion

// top of the file: import { ReactComponent as Menu } from 'src/assets/menu.svg';
      src={Menu}
```

File: src/components/common/RightMainPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
+const RightMainPanel = () => {
+  return (
+    <div className="flex flex-col items-stretch w-[55%] ml-5">
+      <header className="bg-sky-400 flex grow flex-col items-stretch pb-12 pt-12 px-14">
+        <div className="flex-col relative flex min-h-[716px] justify-center items-center ml-4 px-16 py-12">
+          <img
+            loading="lazy"
+            srcSet="src/assets/i6.svg"
```
Comment (eugenialcaraz) [inline src/components/common/RightMainPanel.tsx:8]: Same comment of the imports of the src here

File: src/components/login/LoginForm.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        alt="Image 3"
+      />
+      <div className="text-black text-center text-xs font-semibold tracking-widest self-center mt-5">
+        SIGN UP
+      </div>
+    </div>
+  );
+};
+
+export default LoginForm;
```
Comment (eugenialcaraz) [inline src/components/login/LoginForm.tsx:103]: All the strings in the file manually typed can be converted to constants and reused

File: src/components/signup/SignupForm.tsx [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+import { FormEvent, useState } from "react";
+
+function SignUpForm() {
+  const [firstName, setFirstName] = useState("");
+  const [lastName, setLastName] = useState("");
+  const [username, setUsername] = useState("");
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
+  const [confirmPassword, setConfirm] = useState("");
+  const [gender, setGender] = useState("");
```
Comment (eugenialcaraz) [inline src/components/signup/SignupForm.tsx]:
```suggestion
  const [userData, setUserData] = useState({
  firstName:'',
  lastName:'',
  userName:'',
  email:'',
  password:'',
  confirmPassword:'',
  gender:''});
```

File: src/components/signup/SignupForm.tsx [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    setPassword(e.currentTarget.value);
+  };
+
+  const handleConfirmPasswordChange = (e: FormEvent<HTMLInputElement>) => {
+    setConfirm(e.currentTarget.value);
+  };
+
+  const handleGenderChange = (e: FormEvent<HTMLSelectElement>) => {
+    setGender(e.currentTarget.value);
+  };
```
Comment (eugenialcaraz) [inline src/components/signup/SignupForm.tsx]:
```suggestion
  const handleUserData = (e: FormEvent<HTMLInputElement>, field) => {
    setUserData({...userData, [field]:e.currentTarget.value});
  };
```

File: src/components/signup/SignupForm.tsx [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    e.preventDefault();
+
+    if (
+      firstName === "" ||
+      lastName === "" ||
+      username === "" ||
+      email === "" ||
+      password === "" ||
+      gender === "" ||
+      confirmPassword != password
```
Comment (eugenialcaraz) [inline src/components/signup/SignupForm.tsx]:
```suggestion
     Object.values(userData).some('') ||
      confirmPassword != password
```

File: src/components/signup/SignupForm.tsx [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        <div className="text-black text-center text-xs font-semibold tracking-widest self-center mt-5">
+          <label htmlFor="firstName"> FIRST NAME </label>
+          <input
+            id="firstName"
+            type="text"
+            className="self-stretch flex shrink-0 h-[37px] flex-col mt-2 border-[0.5px] border-solid border-zinc-950 p-3"
+            aria-label="Name"
+            onChange={handleFirstNameChange}
+          />
+        </div>
```
Comment (eugenialcaraz) [inline src/components/signup/SignupForm.tsx]: Maybe creating a common input component will be a good idea here, as you are repeating most of the code

File: src/components/signup/SignupForm.tsx [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+          There was an error registering your account, please try again later.
+        </p>
+        <p
+          className="text-xs text-center text-green-700 font-semibold mt-5"
+          hidden={!registrationSuccess}
+        >
+          REGISTRATION SUCCESS!
+        </p>
+        <hr className="h-0.5 border-none bg-black mt-6 w-2/3 self-center" />
+        <p className="self-center text-xs, mt-3 font-semibold">SIGN IN</p>
```
Comment (eugenialcaraz) [inline src/components/signup/SignupForm.tsx:148]: I think this should be an `<a>` in the future

**Review by eugenialcaraz — COMMENTED** [review]
Hey! Awesome job, Im adding some comments with suggestions and good practices. 
Let me know what do you think!

File: src/components/login/LoginForm.tsx [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    if (!password) {
+      setPasswordError("Please fill in your password");
+    }
+
+    if (email && password) {
+      console.log("Sign in");
+    }
+  };
```
Comment (Ivan-Encinas) [inline src/components/login/LoginForm.tsx:26]: You can avoid all this if you use React hook form, maybe its a good opportunity to try it.

**Review by Ivan-Encinas — CHANGES_REQUESTED** [review]
Great job, just check out the suggestions in the comments

File: src/components/common/MainLayout.tsx (hunk 6 above)
Comment (Rodrigo-Miranda-0) [reply src/components/common/MainLayout.tsx:5]: I actually get the `Property 'children' does not exist on type 'ReactElement<any, string | JSXElementConstructor<any>>'` error once i do this

Comment (Ivan-Encinas) [reply src/components/common/MainLayout.tsx:5]: `const MainLayout = ( children : ReactElement) => ` is without the brackets , you don't have to destructure the children prop

File: src/components/common/LeftMainPanel.tsx (hunk 5 above)
Comment (Ivan-Encinas) [reply src/components/common/LeftMainPanel.tsx:4]: Same here with the brackets in children

**Pull request merged**